File: Source/AlphaTab.JavaScript/Platform/JavaScript/AlphaSynthFlashOutput.cs
using System;
using AlphaTab.Audio.Synth;
using AlphaTab.Audio.Synth.Ds;
using AlphaTab.Collections;
using AlphaTab.Haxe.Js;
using AlphaTab.Haxe.Js.Html;
using Haxe.Js.Html;
using Phase;
using Phase.Attributes;

namespace AlphaTab.Platform.JavaScript
{
    // NOTE: we prefix all ISynthOutput methods with "AlphaSynth" to ensure
    // the ExternalInterface callbacks are called (play, stop etc. might control. the main movie)
    interface IFlashSynthOutput
    {
        void AlphaSynthSequencerFinished();
        void AlphaSynthPlay();
        void AlphaSynthPause();
        void AlphaSynthResetSamples();
        void AlphaSynthAddSamples(string base64Samples);
    }

    class AlphaSynthFlashOutput : ISynthOutput
    {
        public const int PreferredSampleRate = 44100;

        private readonly string _alphaSynthRoot;
        private const string Id = "alphaSynthFlashPlayer";
        private static readonly FastDictionary<string, AlphaSynthFlashOutput> Lookup;

        static AlphaSynthFlashOutput()
        {
            Lookup = new FastDictionary<string, AlphaSynthFlashOutput>();
        }

        private static int NextId;

        private string _id;
        private string _swfId;
        private Element _swfContainer;

        public int SampleRate
        {
            get { return PreferredSampleRate; }
        }

        public AlphaSynthFlashOutput(string alphaSynthRoot)
        {
            _alphaSynthRoot = alphaSynthRoot;

            var lastSlash = _alphaSynthRoot.LastIndexOf("/");
            if (lastSlash != -1)
            {
                _alphaSynthRoot = _alphaSynthRoot.Substring(0, lastSlash + 1);
            }
        }

        public void Open()
        {
            _id = Id + NextId;
            _swfId = _id + "swf";
            Lookup[_id] = this;
            NextId++;

            var document = Browser.Document;
            _swfContainer = document.CreateElement("div");
            _swfContainer.ClassName = Id;
            _swfContainer.SetAttribute("id", _id);
            document.Body.AppendChild(_swfContainer);

            var swf = Lib.Global.swfobject;
            Action<string, string, string, string, string, string, object, object, object> embedSwf = swf.embedSWF;
            embedSwf(
                _alphaSynthRoot + "AlphaSynth.FlashOutput.swf",
                _id, "1px", "1px", "9.0.0",
                null,
                new { id = _id, sampleRate = PreferredSampleRate }, new { allowScriptAccess = "always" }, new { id = _swfId }
            );
        }

        private IFlashSynthOutput FlashOutput
        {
            [Inline]
            get
            {
                var element = Browser.Document.GetElementById(_swfId);
                return Script.Write<IFlashSynthOutput>("untyped __js__(\"{0}\", element)");
            }
        }

        public void Play()
        {
            FlashOutput.AlphaSynthPlay();
        }

        public void Pause()
        {
            FlashOutput.AlphaSynthPause();
        }

        public void SequencerFinished()
        {
            FlashOutput.AlphaSynthSequencerFinished();
        }

        public void AddSamples(SampleArray samples)
        {
            var uint8 = new Uint8Array(samples.ToFloat32Array().Buffer);
            var b64 = Script.Write<string>("untyped __js__(\"window.btoa(String.fromCharCode.apply(null, {0}))\", uint8)");
            FlashOutput.AlphaSynthAddSamples(b64);
        }


        public void ResetSamples()
        {
            FlashOutput.AlphaSynthResetSamples();
        }

        public event Action Ready;
        public static void OnReady(string id)
        {
            if (Lookup.ContainsKey(id))
            {
                Lookup[id].Ready();
            }
        }

        public event Action SampleRequest;
        public static void OnSampleRequest(string id)
        {
            if (Lookup.ContainsKey(id))
            {
                Lookup[id].SampleRequest();
            }
        }

        public event Action Finished;
        public static void OnFinished(string id)
        {
            if (Lookup.ContainsKey(id) && Lookup[id].Finished != null)
            {
                Lookup[id].Finished();
            }
        }

        public event Action<int> SamplesPlayed;
        public static void OnSamplesPlayed(string id, int samples)
        {
            if (Lookup.ContainsKey(id))
            {
                Lookup[id].SamplesPlayed(samples);
            }
        }
    }
}

File: src/platform/javaScript/AlphaSynthFlashOutput.ts
import { ISynthOutput } from '@src/audio/synth/ISynthOutput';
import { EventEmitter } from '@src/EventEmitter';
import { IFlashSynthOutput } from '@src/platform/javaScript/IFlashSynthOutput';
import { Logger } from '@src/util/Logger';

export class AlphaSynthFlashOutput implements ISynthOutput {
    private static readonly Id: string = 'alphaSynthFlashPlayer';
    private static readonly PreferredSampleRate: number = 44100;

    private static Lookup: Map<string, AlphaSynthFlashOutput> = new Map<string, AlphaSynthFlashOutput>();
    private static NextId: number = 0;

    private _alphaSynthRoot: string;
    private _id!: string;
    private _swfId!: string;
    private _swfContainer!: HTMLElement;

    public get sampleRate(): number {
        return AlphaSynthFlashOutput.PreferredSampleRate;
    }

    public constructor(alphaSynthRoot: string) {
        this._alphaSynthRoot = alphaSynthRoot;
        let lastSlash: number = this._alphaSynthRoot.lastIndexOf('/');
        if (lastSlash !== -1) {
            this._alphaSynthRoot = this._alphaSynthRoot.substr(0, lastSlash + 1);
        }
    }

    public open(): void {
        this._id = AlphaSynthFlashOutput.Id + AlphaSynthFlashOutput.NextId;
        this._swfId = this._id + 'swf';
        AlphaSynthFlashOutput.Lookup.set(this._id, this);
        AlphaSynthFlashOutput.NextId++;
        this._swfContainer = document.createElement('div');
        this._swfContainer.className = AlphaSynthFlashOutput.Id;
        this._swfContainer.setAttribute('id', this._id);
        document.body.appendChild(this._swfContainer);
        let swf: any = (window as any)['swfobject'];
        if (swf) {
            swf.embedSWF(
                this._alphaSynthRoot + 'AlphaSynth.FlashOutput.swf',
                this._id,
                '1px',
                '1px',
                '9.0.0',
                null,
                {
                    id: this._id,
                    sampleRate: 44100
                },
                {
                    allowScriptAccess: 'always'
                },
                {
                    id: this._swfId
                }
            );
        } else {
            Logger.error('Player', 'swfobject not found, player will not work', null);
        }
    }

    private get flashOutput(): IFlashSynthOutput {
        let element: HTMLElement | null = document.getElementById(this._swfId);
        return element as any;
    }

    public play(): void {
        this.flashOutput.alphaSynthPlay();
    }

    public pause(): void {
        this.flashOutput.alphaSynthPause();
    }

    public sequencerFinished(): void {
        this.flashOutput.alphaSynthSequencerFinished();
    }

    public addSamples(samples: Float32Array): void {
        let uint8: Uint8Array = new Uint8Array(samples.buffer);
        let b64: string = window.btoa(String.fromCharCode.apply(null, uint8 as any));
        this.flashOutput.alphaSynthAddSamples(b64);
    }

    public resetSamples() {
        this.flashOutput.alphaSynthResetSamples();
    }

    public readonly ready: EventEmitter<() => void> = new EventEmitter();

    public static onReady(id: string): void {
        if (AlphaSynthFlashOutput.Lookup.has(id)) {
            AlphaSynthFlashOutput.Lookup.get(id)!.ready.trigger();
        }
    }

    public readonly samplesPlayed: EventEmitter<(count: number) => void> = new EventEmitter();

    public static onSamplesPlayed(id: string, samples: number): void {
        if (AlphaSynthFlashOutput.Lookup.has(id)) {
            AlphaSynthFlashOutput.Lookup.get(id)!.samplesPlayed.trigger(samples);
        }
    }

    public readonly sampleRequest: EventEmitter<() => void> = new EventEmitter();

    public static onSampleRequest(id: string): void {
        if (AlphaSynthFlashOutput.Lookup.has(id)) {
            AlphaSynthFlashOutput.Lookup.get(id)!.sampleRequest.trigger();
        }
    }

    public readonly finished: EventEmitter<() => void> = new EventEmitter();

    public static onFinished(id: string): void {
        if (AlphaSynthFlashOutput.Lookup.has(id) && AlphaSynthFlashOutput.Lookup.get(id)!.finished) {
            AlphaSynthFlashOutput.Lookup.get(id)!.finished.trigger();
        }
    }

    public activate(): void {}
}
